Mapstruct 1.3.0.Final version is generating wrong code for optional collection getter fields

We have a POJO class with Optional getter fields. The mapstruct 1.3.0.Final generating wrong code for the Optional collection field.

We have ProgramAggregate POJO which contains Collection of Program (Collection) and it is an Optional getter type.

When we run with mapstruct 1.2.0.Final, we are seeing proper code generation. But the same code with 1.3.0.Final generating wrong code. It is not generating the Collection mapping method for Optional Collection getter methods.

![image](https://user-images.githubusercontent.com/47814078/95746732-ed4fab00-0cb4-11eb-925b-fa234370569d.png)

The code generated without a mapping method. 
**Collection<ProgramDto> collection = fromOptional( programAggregate.getPrograms() );  [With Mapstruct 1.3.0.Final]**
It should generate like below :
**Collection<ProgramDto> collection = mapPrograms(fromOptional( programAggregate.getPrograms() ));  [With Mapstruct 1.2.0.Final]**

Note:
**The same code is working fine with map struct 1.2.0.Final version.**

```
@Data
@NoArgsConstructor
@ToString(callSuper = true)
@FieldDefaults(makeFinal = false, level = AccessLevel.PRIVATE)
public class ProgramAggregate {
    public static final long serialVersionUID = 1L;

    Collection<Program> programs;

    public Optional<Collection<Program>> getPrograms() {
        return Optional.ofNullable(programs);
    }


}


@Data
@NoArgsConstructor
@FieldDefaults(makeFinal = false, level = AccessLevel.PRIVATE)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "_type")
public class Program {
    String name;
    String number;


    public Optional<String> getName() {
        return Optional.ofNullable(name);
    }

    public Optional<String> getNumber() {
        return Optional.ofNullable(number);
    }

}


@Data
@NoArgsConstructor
@FieldDefaults(makeFinal = false, level = AccessLevel.PRIVATE)
public class ProgramResponseDto {
    Collection<ProgramDto> programs;

    public Optional<Collection<ProgramDto>> getPrograms() {
        return Optional.ofNullable(programs);
    }

}

@Data
@NoArgsConstructor
@FieldDefaults(makeFinal = false, level = AccessLevel.PRIVATE)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "_type")
public class ProgramDto {
    String name;
    String number;
    String oldNumber;

    public Optional<String> getName() {
        return Optional.ofNullable(name);
    }

    public Optional<String> getNumber() {
        return Optional.ofNullable(number);
    }

    public Optional<String> getOldNumber() {
        return Optional.ofNullable(oldNumber);
    }

}

@Mapper(nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS, unmappedTargetPolicy = ReportingPolicy.WARN,
        collectionMappingStrategy = CollectionMappingStrategy.TARGET_IMMUTABLE)
public interface IProgramMapper extends IOptionalMapper, IDefaultMapper {
    ProgramResponseDto map(ProgramAggregate programAggregate);

    ProgramDto map(Program sourceProgramDto);

    Collection<ProgramDto> mapPrograms(Collection<Program> sourcePrograms);



}
```

Result:
```
@Generated(
    value = "org.mapstruct.ap.MappingProcessor",
    date = "2020-10-12T13:10:32+0530",
    comments = "version: 1.3.0.Final, compiler: javac, environment: Java 1.8.0_231 (Oracle Corporation)"
)
public class IProgramMapperImpl implements IProgramMapper {

    @Override
    public ProgramResponseDto map(ProgramAggregate programAggregate) {
        if ( programAggregate == null ) {
            return null;
        }

        ProgramResponseDto programResponseDto = new ProgramResponseDto();

        **Collection<ProgramDto> collection = fromOptional( programAggregate.getPrograms() );**
        if ( collection != null ) {
            programResponseDto.setPrograms( collection );
        }

        return programResponseDto;
    }

    @Override
    public ProgramDto map(Program sourceProgramDto) {
        if ( sourceProgramDto == null ) {
            return null;
        }

        ProgramDto programDto = new ProgramDto();

        if ( sourceProgramDto.getName() != null ) {
            programDto.setName( fromOptional( sourceProgramDto.getName() ) );
        }
        if ( sourceProgramDto.getNumber() != null ) {
            programDto.setNumber( fromOptional( sourceProgramDto.getNumber() ) );
        }

        return programDto;
    }

    @Override
    public Collection<ProgramDto> mapPrograms(Collection<Program> sourcePrograms) {
        if ( sourcePrograms == null ) {
            return null;
        }

        Collection<ProgramDto> collection = new ArrayList<ProgramDto>( sourcePrograms.size() );
        for ( Program program : sourcePrograms ) {
            collection.add( map( program ) );
        }

        return collection;
    }
}
```
Below is the error after the maven build. Error:

**[ERROR] //mapstruct_latest_example/target/generated-sources/annotations/IProgramMapperImpl.java:[21,57] incompatible types: inference variable T has incompatible bounds [ERROR] equality constraints: java.util.Collection [ERROR] upper bounds: java.util.Collection,java.lang.Object**

Posted the same issue in StackoverFlow:
https://stackoverflow.com/questions/64313831/mapstruct-1-3-0-final-version-is-generating-wrong-code-generation-for-optional-c

--- a/processor/src/main/java/org/mapstruct/ap/internal/model/source/MethodMatcher.java
+++ b/processor/src/main/java/org/mapstruct/ap/internal/model/source/MethodMatcher.java
@@ -296,8 +296,17 @@ public class MethodMatcher {
         public Boolean visitTypeVariable(TypeVariable t, TypeMirror p) {
             if ( genericTypesMap.containsKey( t ) ) {
                 // when already found, the same mapping should apply
+                // Then we should visit the resolved generic type.
+                // Which can potentially be another generic type
+                // e.g.
+                // <T> T fromOptional(Optional<T> optional)
+                // T resolves to Collection<Integer>
+                // We know what T resolves to, so we should treat it as if the method signature was
+                // Collection<Integer> fromOptional(Optional<Collection<Integer> optional)
                 TypeMirror p1 = genericTypesMap.get( t );
-                return typeUtils.isSubtype( p, p1 );
+                // p (Integer) should be a subType of p1 (Number)
+                // i.e. you can assign p (Integer) to p1 (Number)
+                return visit( p, p1 );
             }
             else {
                 // check if types are in bound

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/Issue2233Test.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/Issue2233Test.java
@@ -1,0 +1,50 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2233;
+
+import java.util.Collections;
+import java.util.Optional;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mapstruct.ap.testutil.IssueKey;
+import org.mapstruct.ap.testutil.WithClasses;
+import org.mapstruct.ap.testutil.runner.AnnotationProcessorTestRunner;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.tuple;
+
+/**
+ * @author Filip Hrisafov
+ */
+@IssueKey("2233")
+@RunWith(AnnotationProcessorTestRunner.class)
+@WithClasses( {
+    Program.class,
+    ProgramAggregate.class,
+    ProgramDto.class,
+    ProgramMapper.class,
+    ProgramResponseDto.class,
+} )
+public class Issue2233Test {
+
+    @Test
+    public void shouldCorrectlyMapFromOptionalToCollection() {
+        ProgramResponseDto response = ProgramMapper.INSTANCE.map( new ProgramAggregate( Collections.singleton(
+            new Program(
+                "Optional Mapping",
+                "123"
+            ) ) ) );
+
+        assertThat( response ).isNotNull();
+        assertThat( response.getPrograms() ).isPresent();
+        assertThat( response.getPrograms().get() )
+            .extracting( ProgramDto::getName, ProgramDto::getNumber )
+            .containsExactly(
+                tuple( Optional.of( "Optional Mapping" ), Optional.of( "123" ) )
+            );
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/Program.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/Program.java
@@ -1,0 +1,30 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2233;
+
+import java.util.Optional;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class Program {
+
+    private final String name;
+    private final String number;
+
+    public Program(String name, String number) {
+        this.name = name;
+        this.number = number;
+    }
+
+    public Optional<String> getName() {
+        return Optional.ofNullable( name );
+    }
+
+    public Optional<String> getNumber() {
+        return Optional.ofNullable( number );
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/ProgramAggregate.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/ProgramAggregate.java
@@ -1,0 +1,25 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2233;
+
+import java.util.Collection;
+import java.util.Optional;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class ProgramAggregate {
+
+    private final Collection<Program> programs;
+
+    public ProgramAggregate(Collection<Program> programs) {
+        this.programs = programs;
+    }
+
+    public Optional<Collection<Program>> getPrograms() {
+        return Optional.ofNullable( programs );
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/ProgramDto.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/ProgramDto.java
@@ -1,0 +1,30 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2233;
+
+import java.util.Optional;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class ProgramDto {
+
+    private final String name;
+    private final String number;
+
+    public ProgramDto(String name, String number) {
+        this.name = name;
+        this.number = number;
+    }
+
+    public Optional<String> getName() {
+        return Optional.ofNullable( name );
+    }
+
+    public Optional<String> getNumber() {
+        return Optional.ofNullable( number );
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/ProgramMapper.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/ProgramMapper.java
@@ -1,0 +1,31 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2233;
+
+import java.util.Collection;
+import java.util.Optional;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+/**
+ * @author Filip Hrisafov
+ */
+@Mapper
+public interface ProgramMapper {
+
+    ProgramMapper INSTANCE = Mappers.getMapper( ProgramMapper.class );
+
+    ProgramResponseDto map(ProgramAggregate programAggregate);
+
+    ProgramDto map(Program sourceProgramDto);
+
+    Collection<ProgramDto> mapPrograms(Collection<Program> sourcePrograms);
+
+    default <T> T fromOptional(Optional<T> optional) {
+        return optional.orElse( null );
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/ProgramResponseDto.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2233/ProgramResponseDto.java
@@ -1,0 +1,25 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2233;
+
+import java.util.Collection;
+import java.util.Optional;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class ProgramResponseDto {
+
+    private final Collection<ProgramDto> programs;
+
+    public ProgramResponseDto(Collection<ProgramDto> programs) {
+        this.programs = programs;
+    }
+
+    public Optional<Collection<ProgramDto>> getPrograms() {
+        return Optional.ofNullable( programs );
+    }
+}

